mm2src/coins/utxo/utxo_common.rs

```diff
@@ -58,126 +61,266 @@ lazy_static! {
+    let address = coin.address_from_extended_pubkey(&derived_pubkey);
+
+    let mut derivation_path = hd_account.account_derivation_path.clone();
+    derivation_path.push(change_child);
+    derivation_path.push(address_id_child);
+    Ok(HDAddress {
+        address,
+        derivation_path,
+    })
 }
 
-#[async_trait]
-impl UtxoCoinBuilder for UtxoArcBuilder<'_> {
-    type ResultCoin = UtxoArc;
+pub async fn hd_wallet_balance<T>(coin: &T, hd_wallet: &UtxoHDWallet) -> BalanceResult<Vec<HDAccountBalance>>
+where
+    T: UtxoCommonOps
+        + UtxoHDWalletOps
+        + HDWalletBalanceOps<HDWallet = UtxoHDWallet, HDAccount = UtxoHDAccount, Address = Address>,
+{
+    let orig_accounts = hd_wallet.accounts.lock().clone();
+    let mut accounts = orig_accounts.clone();
+    let mut balances = Vec::with_capacity(accounts.len());
 
-    async fn build(self) -> UtxoCoinBuildResult<Self::ResultCoin> {
-        let utxo = self.build_utxo_fields().await?;
-        Ok(UtxoArc(Arc::new(utxo)))
+    for hd_account in accounts.iter_mut() {
+        balances.push(coin.hd_account_balance(hd_wallet, hd_account).await?);
     }
 
-    fn ctx(&self) -> &MmArc { self.ctx }
+    // Update the accounts container if it hasn't been changed already while we requested the balances.
+    let mut current_accounts = hd_wallet.accounts.lock();
+    if *current_accounts == orig_accounts {
+        *current_accounts = accounts;
+    }
 
-    fn conf(&self) -> &Json { self.conf }
+    Ok(balances)
+}
 
-    fn activation_params(&self) -> UtxoActivationParams { self.activation_params.clone() }
+pub async fn hd_account_balance<T>(
+    coin: &T,
+    hd_wallet: &UtxoHDWallet,
+    hd_account: &mut UtxoHDAccount,
+) -> BalanceResult<HDAccountBalance>
+where
+    T: UtxoCommonOps
+        + UtxoHDWalletOps
+        + HDWalletBalanceOps<HDWallet = UtxoHDWallet, HDAccount = UtxoHDAccount, Address = Address>
+        + Sync,
+{
+    let mut account_balance = HDAccountBalance {
+        account_index: hd_account.account_id as u32,
+        derivation_path: RpcDerivationPath(hd_account.account_derivation_path.clone()),
+        addresses: Vec::new(),
+    };
+
+    let external_addresses = false;
+    let internal_addresses = true;
+
+    // Request balances of the external addresses.
+    account_balance.addresses.extend(
+        balance_of_hd_account_used_addresses(coin, hd_account, external_addresses)
+            .await?
+            .into_iter(),
+    );
+    account_balance.addresses.extend(
+        check_balance_of_hd_account_unknown_addresses(coin, hd_wallet, hd_account, external_addresses)
+            .await?
+            .into_iter(),
+    );
 
-    fn ticker(&self) -> &str { self.ticker }
+    // Request balances of the internal addresses.
+    account_balance.addresses.extend(
+        balance_of_hd_account_used_addresses(coin, hd_account, internal_addresses)
+            .await?
+            .into_iter(),
+    );
+    account_balance.addresses.extend(
```

Comment:
It looks like an overkill to always check the balance of unknown/unused addresses on every balance call. Would it be sufficient to run this check only on coin initialization? We can add a separate RPC to generate a new address when required.

Reply:
Great idea! Will do it

Reply:
Done